`tsc` config picks a binary that does not support `--lsp`, so the server never starts

### Description

Nvim version: `NVIM v0.12.4` (`LuaJIT 2.1.1785763465`, build `RelWithDebInfo`)

Enabled with the stock config shipped in this repo:

```lua
vim.lsp.enable('tsc')
```

`lsp/tsc.lua` resolves its binary in `cmd` using `vim.fn.executable()` only:

```lua
cmd = function(dispatchers, config)
  local cmd = 'tsc'
  local bins = { 'tsc', 'tsgo' }
  for _, bin in ipairs(bins) do
    if (config or {}).root_dir then
      local local_cmd = vim.fs.joinpath(config.root_dir, 'node_modules/.bin', bin)
      if vim.fn.executable(local_cmd) == 1 then
        cmd = local_cmd
        break
      end
    end
    if vim.fn.executable(bin) == 1 then
      cmd = bin
      break
    end
  end
  return vim.lsp.rpc.start({ cmd, '--lsp', '--stdio' }, dispatchers)
end,
```

The `--lsp` flag only exists in the native compiler, TypeScript 7.0 and newer. Any project
that still depends on TypeScript 5.x or 6.x has a `node_modules/.bin/tsc` that is executable
but has no `--lsp` flag at all. Because the project-local path is preferred and only tested
with `executable()`, that binary wins the lookup, gets spawned with `--lsp --stdio`, and
immediately errors out. The client never attaches, and a TypeScript 7.0+ binary that *is*
present on `$PATH` (e.g. installed through Mason) is never considered.

Running the exact command the config builds shows what happens:

```console
$ ./node_modules/.bin/tsc --version
Version 5.9.3

$ ./node_modules/.bin/tsc --lsp --stdio
error TS5023: Unknown compiler option '--lsp'.
error TS5023: Unknown compiler option '--stdio'.
```

The process exits instead of speaking LSP over stdio, so `vim.lsp.rpc.start()` gets a dead
peer. There is no hint in `:checkhealth vim.lsp` that the wrong binary was chosen — the
config looks enabled and the buffer simply has no client.

## Steps to reproduce

1. Have a TypeScript 7.0+ binary on `$PATH` (`tsc --version` → `Version 7.0.2`).
2. Open a project whose `package.json` pins TypeScript 5.x or 6.x, so that
   `node_modules/.bin/tsc` is an older non-native compiler.
3. `nvim src/index.ts`
4. `:lua =vim.lsp.get_clients({ bufnr = 0 })`

## Expected behavior

The config picks a binary that actually supports `--lsp`. If the project-local one is too
old, resolution should fall through to the next candidate (`$PATH` `tsc`, then `tsgo`),
keeping the existing preference order. If no candidate qualifies, the server should not
attach and the user should get a warning explaining why, rather than a silent failure.

## Actual behavior

The project-local pre-7.0 `tsc` is spawned with `--lsp --stdio`, exits with
`error TS5023: Unknown compiler option '--lsp'`, and no client attaches to the buffer.
A working TypeScript 7.0+ binary on `$PATH` is ignored.

## Notes

- The version probe has to happen before `cmd`, otherwise `root_dir` can accept a root for
  which no usable binary exists.
- The probe (`tsc --version`) should be cached per project root so it does not run on every
  client start.

--- a/lsp/tsc.lua
+++ b/lsp/tsc.lua
@@ -52,6 +52,7 @@
 
 local bin_cache = {} ---@type table<string, string>
 
+--- Checks if the given tsc/tsgo cmd supports the "--lsp" arg.
 ---@param bin string
 ---@return boolean
 local function supports_lsp(bin)
@@ -60,9 +61,9 @@ local function supports_lsp(bin)
   end
 
   local out = vim.system({ bin, '--version' }, { text = true }):wait()
-  local major = tonumber((out.stdout or ''):match('(%d+)%.'))
+  local version = vim.version.parse(out.stdout or '')
 
-  return out.code == 0 and major ~= nil and major >= 7
+  return out.code == 0 and version ~= nil and version.major >= 7
 end
 
 ---@type vim.lsp.Config

--- a/lsp/tsc.lua
+++ b/lsp/tsc.lua
@@ -8,6 +8,10 @@
 ---
 --- `tsc` can be installed via npm `npm install typescript`.
 ---
+--- The language server (`--lsp`) is only available in the native compiler, TypeScript 7.0
+--- and newer. An older binary in `node_modules/.bin` is skipped in favour of one on `$PATH`
+--- that does support it; if no candidate qualifies, the server does not attach.
+---
 --- ### Monorepo support
 ---
 --- `tsc` supports monorepos by default. It will automatically find the `tsconfig.json` or `jsconfig.json` corresponding to the package you are working on.
@@ -46,6 +50,21 @@
 --- If DENO ROOT is found, and it's longer than or equal to PROJECT ROOT, then this is a Deno file, and we abort.
 --- Otherwise, attach at PROJECT ROOT, or the cwd if not found.
 
+local bin_cache = {} ---@type table<string, string>
+
+---@param bin string
+---@return boolean
+local function supports_lsp(bin)
+  if vim.fn.executable(bin) ~= 1 then
+    return false
+  end
+
+  local out = vim.system({ bin, '--version' }, { text = true }):wait()
+  local major = tonumber((out.stdout or ''):match('(%d+)%.'))
+
+  return out.code == 0 and major ~= nil and major >= 7
+end
+
 ---@type vim.lsp.Config
 return {
   settings = {
@@ -73,21 +92,7 @@ return {
     },
   },
   cmd = function(dispatchers, config)
-    local cmd = 'tsc'
-    local bins = { 'tsc', 'tsgo' }
-    for _, bin in ipairs(bins) do
-      if (config or {}).root_dir then
-        local local_cmd = vim.fs.joinpath(config.root_dir, 'node_modules/.bin', bin)
-        if vim.fn.executable(local_cmd) == 1 then
-          cmd = local_cmd
-          break
-        end
-      end
-      if vim.fn.executable(bin) == 1 then
-        cmd = bin
-        break
-      end
-    end
+    local cmd = bin_cache[(config or {}).root_dir] or 'tsc'
     return vim.lsp.rpc.start({ cmd, '--lsp', '--stdio' }, dispatchers)
   end,
   filetypes = {
@@ -119,6 +124,26 @@ return {
     end
     -- project is standard TS, not deno
     -- We fallback to the current working directory if no project root is found
-    on_dir(project_root or vim.fn.getcwd())
+    local root = project_root or vim.fn.getcwd()
+
+    if bin_cache[root] then
+      return on_dir(root)
+    end
+
+    local bins = {}
+
+    for _, bin in ipairs({ 'tsc', 'tsgo' }) do
+      bins[#bins + 1] = vim.fs.joinpath(root, 'node_modules/.bin', bin)
+      bins[#bins + 1] = bin
+    end
+
+    for _, bin in ipairs(bins) do
+      if supports_lsp(bin) then
+        bin_cache[root] = bin
+        return on_dir(root)
+      end
+    end
+
+    vim.notify('tsc: no binary supporting `--lsp` found (requires TypeScript 7.0+)', vim.log.levels.WARN)
   end,
 }

